pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -298,62 +383,74 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
 	return err
 }
 
-const (
-	tokenSecretWaitInterval = 20 * time.Millisecond
-	tokenSecretWaitTimes    = 100
-)
+// getTokenSecret return a token secret for a given service account created by this
+// controller or returns nil when there is no token secret created yet.
+func (e *DockercfgController) getTokenSecret(serviceAccount *api.ServiceAccount) *api.Secret {
+	var tokenSecret *api.Secret
+	for _, s := range e.secretsCache.ListKeys() {
+		if strings.HasPrefix(s, serviceAccount.Namespace+"/"+osautil.GetTokenSecretNamePrefix(serviceAccount)) {
+			uncastSecret, _, _ := e.secretsCache.GetByKey(s)
+			if secret := uncastSecret.(*api.Secret); len(secret.Annotations) > 0 {
+				if createdBy, exists := secret.Annotations[api.CreatedByAnnotation]; exists && createdBy == CreateDockercfgSecretsController {
+					tokenSecret = secret
+					break
+				}
+			}
+		}
+	}
+	return tokenSecret
+}
+
+func (e *DockercfgController) hasPendingToken(serviceAccount *api.ServiceAccount) bool {
+	e.pendingTokens.Lock()
+	defer e.pendingTokens.Unlock()
+	return e.pendingTokens.Has(serviceAccount.Namespace + "/" + serviceAccount.Name)
+}
 
 // createTokenSecret creates a token secret for a given service account.  Returns the name of the token
-func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
-	tokenSecret := &api.Secret{
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, bool, error) {
+	if e.hasPendingToken(serviceAccount) {
+		return nil, false, nil
+	}
+	tokenSecret := e.getTokenSecret(serviceAccount)
+	if tokenSecret != nil {
+		return tokenSecret, len(tokenSecret.Data[api.ServiceAccountTokenKey]) > 0, nil
+	}
+	tokenSecret = &api.Secret{
 		ObjectMeta: api.ObjectMeta{
 			Name:      secret.Strategy.GenerateName(osautil.GetTokenSecretNamePrefix(serviceAccount)),
```

Comment:
Why not change this to specify a complete name so that you can update a `pendingToken` cache *before* creation, so that you can read from the `pendingTokens` on secret add/update to then re-queue the related SA.  An inspection of the SA, plus available cached secrets ought to allow you to continue from that state.